### PR TITLE
fix(http): reject invalid Host authorities

### PR DESCRIPTION
**Pull request by nathanwhit**
## Summary

- validate HTTP/1 `Host` field values before request dispatch
- reject malformed and repeated `Host` fields consistently across `Deno.serve`, `Deno.serveHttp`, and `node:http`
- keep the existing missing or empty `Host` fallback and leave other field-value handling unchanged

## Details

HTTP/1 `Host` field values use URI host syntax with an optional numeric port. Previously, non-ASCII and malformed authority values could reach request callbacks, and the server paths did not handle all inputs consistently. This change returns a 400 response before invoking callbacks or sending `100 Continue`.

## Tests

- `cargo test -p deno_http_h1`
- `cargo clippy -p deno_http_h1 -p deno_http -p deno_node --features deno_core/v8 -- -D warnings`
- `cargo test -p specs_tests --test specs -- http_host_authority`
- `deno lint tests/specs/run/http_host_authority/main.ts`
- `./tools/format.js <changed files>`

### CODE REVIEW

File: ext/http/lib.rs [hunk 1]
```diff
@@ -69,10 +69,12 @@ use hyper::header::CONTENT_ENCODING;
 use hyper::header::CONTENT_LENGTH;
 use hyper::header::CONTENT_RANGE;
 use hyper::header::CONTENT_TYPE;
+use hyper::header::CONNECTION;
```
Comment (bartlomieju) [inline ext/http/lib.rs]: `rustfmt` reorders these: `CONNECTION` sorts before `CONTENT_ENCODING` (`CONN` < `CONT`), and `HOST` on line 77 sorts before `HeaderMap` (ASCII `O` = 79 < `e` = 101). Along with the width issue on line 960, this means `./tools/format.js` wasn't actually run over `ext/http/lib.rs`, despite the PR description listing it.

File: ext/http/lib.rs [hunk 2]
```diff
@@ -955,6 +957,17 @@ impl Service<Request<Incoming>> for HttpService {
   >;
 
   fn call(&self, request: Request<Incoming>) -> Self::Future {
+    if matches!(request.version(), http::Version::HTTP_10 | http::Version::HTTP_11)
```
Comment (bartlomieju) [inline ext/http/lib.rs]: 83 columns; `.rustfmt.toml` sets `max_width = 80`. CI's `./tools/format.js --check` will fail here.

Running `rustfmt --check` over the branch also turns up diffs at `libs/http_h1/parse.rs:466`, `:520`, `:871` and `ext/node/ops/llhttp/binding.rs:424` — all wrapped expressions that fit on one line. A single `./tools/format.js` pass over the changed files clears all of them.

File: libs/http_h1/parse.rs [hunk 3]
```diff
@@ -368,7 +376,7 @@ impl HeaderInfo {
       0 if body_kind == BodyKind::Empty => Ok(()),
       0 => Err(ParseError::MissingHost),
       1 => Ok(()),
-      _ => Err(ParseError::MultipleHost),
+      _ => unreachable!(),
```
Comment (bartlomieju) [inline libs/http_h1/parse.rs]: I'd rather not have an `unreachable!()` on a path that processes untrusted input, even a provably-correct one.

It only holds because of the `host_count > 1` early return ~10 lines up — a non-local invariant, and specifically one this PR just *moved* out from under the version gate. If a later change moves it back (which is exactly where it lived before), a repeated `Host` header reaches this arm, and release/bench profiles are `panic = "abort"` (`Cargo.toml:487`) — so a remote process abort, in the hardening patch.

`_ => Err(ParseError::MultipleHost)` costs nothing and stays correct regardless of where the early return sits.

File: ext/node/ops/llhttp/binding.rs [hunk 4]
```diff
@@ -412,6 +416,16 @@ unsafe extern "C" fn on_header_value_complete(
...
     .map_or(0, |p| p + 1);
   value.truncate(trimmed_len);
+  if unsafe { (*parser).type_ } == sys::HTTP_REQUEST as u8
+    && field.eq_ignore_ascii_case(b"host")
+  {
+    inner.host_count += 1;
+    if inner.host_count > 1
+      || !deno_http_h1::is_valid_host_header_value(&value)
+    {
+      return -1;
```
Comment (bartlomieju) [inline ext/node/ops/llhttp/binding.rs]: The 400 goes out correctly, but the diagnostic that reaches user code is unhelpful. llhttp collapses this `-1` into `state->error = 0x1d` with `reason = "`on_header_value_complete` callback error"` (`c/llhttp.c:7970`), and `http_parser.ts:242` turns that into `err.code = "HPE_CB_HEADER_VALUE_COMPLETE"` / `err.message = "Parse Error: `on_header_value_complete` callback error"`. That's what a user's `clientError` handler sees — an internal callback name, with nothing indicating the `Host` header was the problem.

This file already has the mechanism: the JS-exception path (~line 690) calls `sys::llhttp_set_error_reason(parser, c"HPE_JS_EXCEPTION:JS Exception")` before returning. Same thing here with e.g. `c"HPE_INVALID_HEADER_TOKEN:Invalid Host header"` would give `err.reason` something actionable.

File: libs/http_h1/parse.rs [hunk 5]
```diff
@@ -418,6 +426,105 @@ impl HeaderInfo {
   }
 }
 
+/// Returns whether a `Host` field value matches the URI host and optional
+/// port grammar used by HTTP/1.
+pub fn is_valid_host_header_value(value: &[u8]) -> bool {
+  if value.is_empty() {
+    return true;
+  }
+  if !value.is_ascii() {
```
Comment (bartlomieju) [inline libs/http_h1/parse.rs:435]: Worth calling out in the PR body / release notes: this is a user-visible break, not just a hardening no-op.

`valid_header_value` permits non-ASCII generally — your own spec test asserts `X-Byte-Value: 0xc3 0xbc` → 200 with the handler observing `"Ã¼"`. So raw-UTF-8 IDN hosts like `Host: ü.com` previously reached handlers, with the WHATWG URL parser punycoding them; now they're a 400. Rejecting is spec-correct and the real-world impact is low (browsers send punycode), but it should be a deliberate, documented change rather than a side effect.

Same goes for zone-ID literals — `[fe80::1%25eth0]` (RFC 6874) is now rejected, and the unit test asserts it.

File: libs/http_h1/parse.rs [hunk 6]
```diff
@@ -418,6 +426,105 @@ impl HeaderInfo {
...
+  }
+
+  if value[0] == b'[' {
+    return is_valid_ip_literal_authority(value);
+  }
+
+  let (host, port) = match value.iter().position(|&byte| byte == b':') {
+    Some(index) => {
+      let port = &value[index + 1..];
+      if port.contains(&b':') {
```
Comment (bartlomieju) [inline libs/http_h1/parse.rs]: This is dead — `port.iter().all(u8::is_ascii_digit)` at the bottom of the function already rejects a `:` in the port. Can just drop the branch and let `(&value[..index], Some(port))` fall through.

File: libs/http_h1/parse.rs [hunk 7]
```diff
@@ -714,6 +821,68 @@ mod tests {
...
+      parse_request_head(
+        b"GET / HTTP/1.0\r\nHost: example.com\r\nHost: example.org\r\n\r\n",
+        &mut headers,
+      ),
+      Err(ParseError::MultipleHost)
+    );
+  }
+
+  #[test]
+  fn validates_host_header_values() {
```
Comment (bartlomieju) [inline libs/http_h1/parse.rs:829]: Two cases worth adding to the table, both currently accepted and both looking like consequences of `all()` on an empty slice rather than deliberate choices:

- empty port — `example.com:` and `[::1]:` (the bracketed branch reaches the same outcome via `suffix[1..]` being empty)
- `[v1.example]:8080` — IPvFuture *with* a port; only the bare form is covered

If empty-port acceptance is intended, pinning it here documents that.

File: ext/node/Cargo.toml [hunk 8]
```diff
@@ -26,6 +26,7 @@ deno_core.workspace = true
 deno_dotenv.workspace = true
 deno_error.workspace = true
 deno_fs.workspace = true
+deno_http_h1.workspace = true
```
Comment (bartlomieju) [inline ext/node/Cargo.toml]: Minor: `deno_node` picks up a whole `deno_http_h1` dependency for one ~100-line pure function with no HTTP state in it. Not a blocker, but if this predicate grows more callers it'd be worth a smaller shared crate rather than widening this edge.

File: tests/specs/run/http_host_authority/main.ts [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    await send(
+      port,
+      request(
+        new Uint8Array([0xc3, 0xbc, 0x2e, 0x63, 0x6f, 0x6d]),
+        encoder.encode("Expect: 100-continue\r\nContent-Length: 1\r\n"),
+      ),
+    ),
+    400,
+  );
+  assertEquals(calls, 9);
```
Comment (bartlomieju) [inline tests/specs/run/http_host_authority/main.ts]: Brittle — any case added above silently requires bumping this magic number, and getting it wrong fails in a way that doesn't point at the cause. Counting the expected-200 cases (or asserting per-case that `calls` incremented / didn't) would survive edits.

Also: this test covers empty `Host:` for `Deno.serve`, but not for `node:http` or `Deno.serveHttp`. Since preserving the empty/missing-`Host` fallback is an explicit goal of the PR, it'd be good to pin it on all three paths.

**Review by bartlomieju — COMMENTED** [review]
Approach looks right to me. Validating before the callback and before hyper can emit `100 Continue` is the correct placement, and covering all three server stacks (`serve_http11_raw` → `libs/http_h1`, hyper's `HttpService`, and llhttp) from a single predicate is the right factoring. The validator itself is careful — bracketed IP-literal handling, IPvFuture, percent-encoding in reg-name, and scanning for the first `]` so `[::1]]` is rejected.

I did check whether the `-1` return from the llhttp callback could reach `llhttp_errno_name`'s `default: abort()` (`c/api.c:212`), since that would have been a remote abort. It can't — llhttp normalizes any non-zero callback return to `state->error = 0x1d` before it's stored (`c/llhttp.c:7970`). No new exposure found there.

Two things I'd want fixed in this pass — formatting (CI will fail) and the `unreachable!()`. The rest are notes.

One item that doesn't map to a line: **HTTP/2 isn't covered.** The hyper check is gated on `HTTP_10 | HTTP_11` and `libs/http_h1` is h1-only, so a `Host` header sent over h2 (`serve_http2_unconditional`) still reaches handlers unvalidated. Probably fine in practice since h2 routes on `:authority`, but the PR description says "consistently across `Deno.serve`, `Deno.serveHttp`, and `node:http`" and that consistency stops at the protocol boundary — worth either covering or saying it's out of scope.

**Review by bartlomieju — CHANGES_REQUESTED** [review]
The `Deno.serve` half of this is good and the grammar implementation is careful. The `node:http` half contradicts documented Node behavior, and node_compat catches it on every platform.

**Blocking: `test-http-server-multiheaders.js` fails on all five platforms.**

That's one of Node's own tests, and its whole point is duplicate-header handling. It sends three `Host` headers and asserts the server *accepts* the request:

```js
headers: [
  ['host', 'foo'],
  ['Host', 'bar'],
  ['hOst', 'baz'],
  ...
]
```
```js
assert.strictEqual(req.headers.host, 'foo');
```

Node keeps the first `Host` and silently drops the rest — `host` is on `_addHeaderLine`'s discard-duplicates list, alongside `content-type`, `referer` and friends. The `host_count > 1 → return -1` in `on_header_value_complete` makes Deno reject the request instead, so `node:http` stops matching Node on a case Node has a regression test for.

I'd drop the duplicate-`Host` rejection from the llhttp path entirely and keep first-wins there. RFC 9112's "MUST respond with 400" is a requirement on an HTTP/1.1 origin server, and `Deno.serve` is the right place to enforce it; `node:http` is emulating a different implementation whose observable behavior is fixed by its test suite. The same reasoning applies to `is_valid_host_header_value` on that path — llhttp is deliberately lenient about field values, and tightening it there will surface as more compat divergence, just not covered by a test yet.

Also worth deciding deliberately: is `-1` from `on_header_value_complete` even reaching a 400? The summary says the change "returns a 400 response before invoking callbacks", which is clearly true for the `h1` and `serveHttp` paths, but aborting the llhttp parse looks more like a parser error than a response. If the connection just gets destroyed, that diverges from Node again — Node writes `400 Bad Request` and closes.

**Also failing: lint.** `tests/specs/run/http_host_authority/main.ts` isn't formatted — two `for (const host of [` lines need re-indenting. `./tools/format.js` fixes it.

**The parts I'm happy with:**

- `is_valid_host_header_value` tracks RFC 3986's `host = IP-literal / IPv4address / reg-name` accurately. `is_valid_reg_name` accepts unreserved + sub-delims + pct-encoded with the correct two-hex-digit lookahead, and IPv4 falls out as a subset of reg-name rather than needing its own branch. Delegating the bracketed case to `Ipv6Addr::parse` after peeling `[`/`]` is the right move, and handling `IPvFuture` rather than rejecting it is more care than I expected.
- Requiring the suffix after `]` to be exactly `:` followed by digits closes `[::1]garbage`.
- Rejecting non-ASCII is correct — `Host` is an authority, so an IDN has to arrive as punycode.

**One deliberate-decision question on the `h1` path:** moving the `host_count > 1` check above the `version != Version::Http11` early return means HTTP/1.0 requests with duplicate `Host` are now rejected too. RFC 9112 scopes that MUST to HTTP/1.1. `rejects_multiple_http10_hosts` says it's intentional, and I don't object — just flagging that it's a behavior change beyond what the RFC requires, so it belongs in the summary.

Happy to re-review once node_compat is green.
